pkg/labels/labels.go

```diff
@@ -131,21 +131,23 @@ func (ls Labels) Hash() uint64 {
 }
 
 // HashForLabels returns a hash value for the labels matching the provided names.
-func (ls Labels) HashForLabels(names ...string) uint64 {
-	b := make([]byte, 0, 1024)
-
-	for _, v := range ls {
-		for _, n := range names {
-			if v.Name == n {
-				b = append(b, v.Name...)
-				b = append(b, sep)
-				b = append(b, v.Value...)
-				b = append(b, sep)
-				break
-			}
+func (ls Labels) HashForLabels(b []byte, names ...string) (uint64, []byte) {
+	b = b[:0]
+	i, j, n1, n2 := 0, 0, len(names), len(ls)
+	for i < n1 {
+		j = sort.Search(n2, func(idx int) bool {
+			return ls[idx].Name >= names[i]
+		})
+
+		if j < n2 && ls[j].Name == names[i] {
+			b = append(b, ls[j].Name...)
+			b = append(b, sep)
+			b = append(b, ls[j].Value...)
+			b = append(b, sep)
 		}
+		i++
 	}
-	return xxhash.Sum64(b)
+	return xxhash.Sum64(b), b
```

Comment:
Why return b here?

Reply:
If the slice was reallocated in `append()` if the size was not enough, the resized slice can be returned passed in the next call.

Reply:
Oh I see you want to preserve any extensions that have happened to the buffer.  SGTM